apps/anoma_node/lib/node/transaction/shard_supervisor.ex

```diff
@@ -0,0 +1,256 @@
+  initial value.
+  """
+  @type schema_t :: [key_t() | {key_t(), initial_value_t()}]
+
+  @typedoc """
+  I am the sharding strategy.
+  Currently, I only support `:one_per_key`.
+  """
+  @type strategy_t :: :one_per_key
+
+  @typedoc """
+  I am the type of the arguments that the ShardSupervisor expects at startup.
+  I require `:node_id` and optionally `:strategy` and `:schema` keys.
+  """
+  @type supervisor_args_t :: [
+          node_id: String.t(),
+          strategy: strategy_t() | nil,
+          schema: schema_t() | nil
+        ]
+
+  @typedoc """
+  I am the type of the arguments that the Shard process expects.
+  I am not explicitly used, but this may be useful to know.
+  """
+  @type shard_args_t :: [
+          id: key_t(),
+          initial_kv: %{key_t() => initial_value_t()}
+        ]
+
+  ############################################################
+  #                 Supervisor Implementation                #
+  ############################################################
+
+  @doc """
+  I am the start_link function for the ShardSupervisor.
+
+  I start and link the supervisor process under the current supervision tree,
+  registering myself locally using a node-specific name.
+  """
+  @spec start_link(args :: supervisor_args_t()) :: Supervisor.on_start()
+  def start_link(args) do
+    name = Registry.via(args[:node_id], __MODULE__)
+    Supervisor.start_link(__MODULE__, args, name: name)
+  end
+
+  @impl true
+  @doc """
+  I am the Supervisor initialization callback.
+
+  I set the process label. If valid :strategy and :schema are provided,
+  I calculate the full key->name mapping and shard child specs,
+  populate a named ETS table with key -> shard ID mappings, and
+  start all configured `Shard` children using a `:one_for_one` strategy.
+  """
+  @spec init(args :: supervisor_args_t()) ::
+          {:ok, {Supervisor.sup_flags(), [Supervisor.child_spec()]}}
+  def init(args) do
+    node_id = Keyword.fetch!(args, :node_id)
+    Process.set_label({__MODULE__, node_id})
+
+    try do
+      # Process schema only if strategy and schema are validly provided
+      {shard_child_specs, key_to_id_map} =
+        process_schema(
+          node_id,
+          Keyword.get(args, :strategy),
+          Keyword.get(args, :schema)
+        )
+
+      # Create ETS table if shards were generated
+      case create_and_populate_ets_table(key_to_id_map, node_id) do
+        {:ok, _ets_tid} ->
+          # Simply initialize the supervisor with the child specs
+          Supervisor.init(shard_child_specs, strategy: :one_for_one)
+
+        :no_shards ->
+          # No shards configured or generated, start no children
+          Supervisor.init([], strategy: :one_for_one)
+
+        {:error, _reason} ->
+          {:stop, :ets_table_creation_failed}
+      end
+    rescue
+      e ->
+        # Reraise the exception to ensure supervisor termination/restart according to strategy
+        reraise e, __STACKTRACE__
+    end
```

Comment:
What's the point of the `try do`?

Just run the code if it crashes it'll already cause the supervisor above this to fail